### PR TITLE
Alistairbntl/lsc

### PR DESCRIPTION
**Pull request by alistairbntl**
Incorporates the LSC preconditioning class.

### CODE REVIEW

**Comment by alistairbntl** [comment]
This pull request is not nearly as large as the number of new lines would suggest.  Most of those lines are for tests.

File: proteus/LinearAlgebraTools.py [hunk 1]
```diff
@@ -153,11 +176,36 @@ def superlu_2_petsc4py(sparse_superlu):
...
+    diag_inv.setUp()
+    diag_inv.setDiagonal(1./sparse_petsc.getDiagonal())
+    return diag_inv
+
 class ParVec:
     """
     A parallel vector built on top of daetk's wrappers for petsc
     """
-    def __init__(self,array,blockSize,n,N,nghosts=None,subdomain2global=None,blockVecType="simple"):#"block"
+    def __init__(self,
```
Comment (adimako) [inline proteus/LinearAlgebraTools.py]: Add documentation for this one?

File: proteus/LinearAlgebraTools.py [hunk 2]
```diff
@@ -518,6 +566,94 @@ def _converged_trueRes(self,ksp,its,rnorm):
...
+        convergenceTest = 'r-true'
+        if convergenceTest == 'r-true':
+            self.r_work = self.BQinvBt.getVecLeft()
+            self.rnorm0 = None
+            self.kspBQinvBt.setConvergenceTest(self._converged_trueRes)
+        else:
+            self.r_work = None        
+        self.kspBQinvBt.setUp()
+
+    def apply(self,A,x,y):
```
Comment (adimako) [inline proteus/LinearAlgebraTools.py]: Documentation?

File: proteus/LinearSolvers.py [hunk 3]
```diff
@@ -718,6 +722,28 @@ def __init__(self, linear_smoother, pde_type='general_saddle_point'):
...
 
+    def getQv(self, output_matrix=False, recalculate=False):
+        """ Return the pressure mass matrix Qp.
+
+        Parameters
+        ----------
+        output_matrix : bool 
+            Determines whether matrix should be exported.
+        recalculate : bool
+            Flag indicating whther matrix should be rebuilt every time it's used
```
Comment (adimako) [inline proteus/LinearSolvers.py]: replace whther with wether

File: proteus/LinearAlgebraTools.py (hunk 1 above)
Comment (alistairbntl) [inline proteus/LinearAlgebraTools.py]: I'm going to leave this documentation for a future PR.

**Pull request merged**